test.js

```diff
@@ -0,0 +1,193 @@
+  dockerCleanup(function(cleanupErr) {
+    if (cleanupErr) {
+      console.error('error cleaning up:', cleanupErr);
+    }
+    if (testErr) {
+      console.error('error running tests:', testErr);
+      process.exit(1);
+    }
+  });
+});
+
+function sleep(n) {
+  return function delayedPassThrough() {
+    var args = [].slice.call(arguments);
+    // last argument is the callback
+    var next = args.pop();
+    // prepend `null` to indicate no error
+    args.unshift(null);
+    setTimeout(function() {
+      next.apply(null, args);
+    }, n);
+  };
+}
+
+function dockerStart(imgName) {
+  return function pullAndStart(next) {
+    console.log('pulling image: %s', imgName);
+    docker.pull(imgName, function(err, stream) {
+      docker.modem.followProgress(stream, function(err, output) {
+        if (err) {
+          return next(err);
+        }
+        console.log('starting container from image: %s', imgName);
+        docker.createContainer({
+          Image: imgName,
+          HostConfig: {
+            PublishAllPorts: true,
+          },
+        }, function(err, container) {
+          console.log('recording container for later cleanup: ', container.id);
+          containerToDelete = container;
+          if (err) {
+            return next(err);
+          }
+          container.start(function(err, data) {
+            next(err, container);
+          });
+        });
+      });
+    });
+  };
+}
+
+function setCloudantEnv(container, next) {
+  container.inspect(function(err, c) {
+    // if swarm, Node.Ip will be set to actual node's IP
+    // if not swarm, but remote docker, use docker host's IP
+    // if local docker, use localhost
+    var host = _.get(c, 'Node.IP', _.get(docker, 'modem.host', '127.0.0.1'));
+    // container's port 80 is dynamically mapped to an external port
+    var port = _.get(c,
+      ['NetworkSettings', 'Ports', '80/tcp', '0', 'HostPort']);
+    process.env.CLOUDANT_PORT = port;
+    process.env.CLOUDANT_HOST = host;
+    process.env.CLOUDANT_URL = 'http://' + host + ':' + port;
+    console.log('env:', _.pick(process.env, [
+      'CLOUDANT_URL',
+      'CLOUDANT_HOST',
+      'CLOUDANT_PORT',
+      'CLOUDANT_USERNAME',
+      'CLOUDANT_PASSWORD',
+      'CLOUDANT_DATABASE',
+    ]));
+    next(null, container);
+  });
+}
+
+function waitFor(path) {
+  return function waitForPath(container, next) {
+    var opts = {
+      host: process.env.CLOUDANT_HOST,
+      port: process.env.CLOUDANT_PORT,
+      auth: process.env.CLOUDANT_USERNAME + ':' + process.env.CLOUDANT_PASSWORD,
+      path: path,
+    };
+
+    console.log('waiting for instance to respond');
+    return setImmediate(ping, null, 30);
+
+    function ping(err, tries) {
+      console.log('ping (%d/30)', 30 - tries);
+      if (tries < 1) {
+        next(err || new Error('failed to contact Cloudant'));
+      }
+      http.get(opts, function(res) {
+        res.on('error', tryAgain);
+        res.on('data', function() {});
```

Comment:
The no-op `'data'` listener is there to consume the stream so that it ends. I'll try to make that more clear.